### PR TITLE
drivers: display: st7789v: Simplify configuration code

### PR DESCRIPTION
**Pull request by caspermeijn**
In preparation of another change I simplified the code for configuration of the st7789v driver. In several steps almost all the data coming from the device tree moved to the new `struct st7789v_config`.  It is const so that it is stored in FLASH.
I also made naming more consistent, made software reset usable and fix compiling with power management.

I tested this with `samples/drivers/display` on `pinetime_devkit0`.

Before:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       25712 B       512 KB      4.90%
            SRAM:       20744 B        64 KB     31.65%
        IDT_LIST:          0 GB         2 KB      0.00%
```

After:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       25756 B       512 KB      4.91%
            SRAM:       20680 B        64 KB     31.56%
        IDT_LIST:          0 GB         2 KB      0.00%
```

### CODE REVIEW

**Comment by caspermeijn** [comment]
I think the failed test is not from my code changes. I will rebase tomorrow and try again.

File: drivers/display/display_st7789v.c [hunk 1]
```diff
@@ -101,16 +99,11 @@ static void st7789v_exit_sleep(struct st7789v_data *data)
...
 {
 	LOG_DBG("Resetting display");
-#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
 	k_sleep(K_MSEC(1));
 	gpio_pin_set(data->reset_gpio, ST7789V_RESET_PIN, 1);
 	k_sleep(K_MSEC(6));
 	gpio_pin_set(data->reset_gpio, ST7789V_RESET_PIN, 0);
 	k_sleep(K_MSEC(20));
-#else
-	st7789v_transmit(p_st7789v, ST7789V_CMD_SW_RESET, NULL, 0);
```
Comment (jfischer-no) [inline drivers/display/display_st7789v.c:111]: This software reset does not work or what is reasoning?

File: drivers/display/display_st7789v.c [hunk 2]
```diff
@@ -79,57 +72,59 @@ static void st7789v_set_lcd_margins(struct st7789v_data *data,
...
-static void st7789v_set_cmd(struct st7789v_data *data, int is_cmd)
+static void st7789v_set_cmd(struct st7789v_config *config, int is_cmd)
 {
-	gpio_pin_set(data->cmd_data_gpio, ST7789V_CMD_DATA_PIN, is_cmd);
+	gpio_pin_set_dt(&config->cmd_data_gpio, is_cmd);
 }
 
-static void st7789v_transmit(struct st7789v_data *data, uint8_t cmd,
-		uint8_t *tx_data, size_t tx_count)
+static void st7789v_transmit(struct st7789v_data *data, struct st7789v_config *config, uint8_t cmd,
```
Comment (jfischer-no) [inline drivers/display/display_st7789v.c]: No, please use `const struct device *dev`
```
static void st7789v_transmit(const struct device *dev, uint8_t cmd,
                             uint8_t *tx_data, size_t tx_count)
```

Same for st7789v_exit_sleep, t7789v_set_mem_area, ...

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/display/display_st7789v.c [hunk 3]
```diff
@@ -64,68 +61,63 @@ struct st7789v_data {
...
 {
 	LOG_DBG("Resetting display");
-#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
 	k_sleep(K_MSEC(1));
-	gpio_pin_set(data->reset_gpio, ST7789V_RESET_PIN, 1);
+	gpio_pin_set_dt(&config->reset_gpio, 1);
 	k_sleep(K_MSEC(6));
-	gpio_pin_set(data->reset_gpio, ST7789V_RESET_PIN, 0);
+	gpio_pin_set_dt(&config->reset_gpio, 0);
 	k_sleep(K_MSEC(20));
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:105]:
```c
if (config->reset_gpio.port != NULL) {
     /* hw reset */
} else {
     /* sw reset */
}
```

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/display/display_st7789v.c (hunk 1 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c:111]: @jfischer-no Well, on `main` branch it is currently not possible to enable software reset. To enable software reset you need to leave out the reset pin, however defining a reset pin is required. This indicated to me that nobody needed software reset and that is was never tested. I thought this change would simplify the code.

File: drivers/display/display_st7789v.c (hunk 2 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c]: I thought this version is more type-safe. I will change to `device *`.

File: drivers/display/display_st7789v.c [hunk 4]
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
...
+
+	gpio_pin_set_dt(&config->cmd_data_gpio, is_cmd);
 }
 
-static void st7789v_transmit(struct st7789v_data *data, uint8_t cmd,
-		uint8_t *tx_data, size_t tx_count)
+static void st7789v_transmit(const struct device *dev, uint8_t cmd,
+			     uint8_t *tx_data, size_t tx_count)
 {
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:82]:
```suggestion
	struct st7789v_config *config = dev->config;
```

File: drivers/display/display_st7789v.c [hunk 5]
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
...
+
 	data->x_offset = x_offset;
 	data->y_offset = y_offset;
 }
 
-static void st7789v_set_cmd(struct st7789v_data *data, int is_cmd)
+static void st7789v_set_cmd(const struct device *dev, int is_cmd)
 {
-	gpio_pin_set(data->cmd_data_gpio, ST7789V_CMD_DATA_PIN, is_cmd);
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:74]:
```suggestion
	struct st7789v_config *config = dev->config;
```

File: drivers/display/display_st7789v.c [hunk 6]
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
 #define ST7789V_PIXEL_SIZE 3u
 #endif
 
-static void st7789v_set_lcd_margins(struct st7789v_data *data,
-			     uint16_t x_offset, uint16_t y_offset)
+static void st7789v_set_lcd_margins(const struct device *dev,
+				    uint16_t x_offset, uint16_t y_offset)
 {
+	struct st7789v_data *data = (struct st7789v_data *)dev->data;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:66]:
```suggestion
	struct st7789v_data *data = dev->data;
```

File: drivers/display/display_st7789v.c [hunk 7]
```diff
@@ -63,69 +60,68 @@ struct st7789v_data {
...
 {
-	st7789v_transmit(data, ST7789V_CMD_SLEEP_OUT, NULL, 0);
+	st7789v_transmit(dev, ST7789V_CMD_SLEEP_OUT, NULL, 0);
 	k_sleep(K_MSEC(120));
 }
 
-static void st7789v_reset_display(struct st7789v_data *data)
+static void st7789v_reset_display(const struct device *dev)
 {
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:106]:
```suggestion
	struct st7789v_config *config = dev->config;
```

File: drivers/display/display_st7789v.c [hunk 8]
```diff
@@ -138,21 +134,24 @@ static int st7789v_read(const struct device *dev,
 	return -ENOTSUP;
 }
 
-static void st7789v_set_mem_area(struct st7789v_data *data, const uint16_t x,
-				 const uint16_t y, const uint16_t w, const uint16_t h)
+static void st7789v_set_mem_area(const struct device *dev,
+				 const uint16_t x, const uint16_t y,
+				 const uint16_t w, const uint16_t h)
 {
+	struct st7789v_data *data = (struct st7789v_data *)dev->data;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:141]:
```suggestion
	struct st7789v_data *data = dev->data;
```

File: drivers/display/display_st7789v.c [hunk 9]
```diff
@@ -161,7 +160,7 @@ static int st7789v_write(const struct device *dev,
 			 const struct display_buffer_descriptor *desc,
 			 const void *buf)
 {
-	struct st7789v_data *data = (struct st7789v_data *)dev->data;
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:163]:
```suggestion
	struct st7789v_config *config = dev->config;
```

File: drivers/display/display_st7789v.c [hunk 10]
```diff
@@ -223,11 +222,11 @@ static int st7789v_set_contrast(const struct device *dev,
 static void st7789v_get_capabilities(const struct device *dev,
 			      struct display_capabilities *capabilities)
 {
-	struct st7789v_data *data = (struct st7789v_data *)dev->data;
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:225]:
```suggestion
	struct st7789v_config *config = dev->config;
```

File: drivers/display/display_st7789v.c [hunk 11]
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
 	return -ENOTSUP;
 }
 
-static void st7789v_lcd_init(struct st7789v_data *p_st7789v)
+static void st7789v_lcd_init(const struct device *dev)
 {
+	struct st7789v_data *data = (struct st7789v_data *)dev->data;
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:268]:
```suggestion
	struct st7789v_data *data = dev->data;
	struct st7789v_config *config = dev->config;
```

File: drivers/display/display_st7789v.c (hunk 1 above)
Comment (gmarull) [reply drivers/display/display_st7789v.c:111]: @caspermeijn can you remove the `required` field from the reset-gpios? Then you can obtain the gpio using `GPIO_DT_SPEC_INST_GET_OR(0, reset_gpios, {0})`. This way you can then add both, hard and soft reset options.

File: drivers/display/display_st7789v.c [hunk 12]
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
...
 	struct st7789v_data *data = (struct st7789v_data *)dev->data;
+	struct st7789v_config *config = (struct st7789v_config *)dev->config;
 
-	data->spi_dev = device_get_binding(DT_INST_BUS_LABEL(0));
-	if (data->spi_dev == NULL) {
-		LOG_ERR("Could not get SPI device for LCD");
-		return -EPERM;
+	if (!spi_is_ready(&config->spi)) {
+		LOG_ERR("SPI device for LCD is not ready");
+		return -EINVAL;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:337]:
```suggestion
		return -ENODEV;
```

File: drivers/display/display_st7789v.c [hunk 13]
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
...
-
-#if DT_INST_NODE_HAS_PROP(0, reset_gpios)
-	data->reset_gpio = device_get_binding(
-			DT_INST_GPIO_LABEL(0, reset_gpios));
-	if (data->reset_gpio == NULL) {
-		LOG_ERR("Could not get GPIO port for display reset");
-		return -EPERM;
+	if (!device_is_ready(config->reset_gpio.port)) {
+		LOG_ERR("GPIO port for display reset is not ready");
+		return -EINVAL;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:342]:
```suggestion
		return -ENODEV;
```

File: drivers/display/display_st7789v.c [hunk 14]
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
...
 #endif
 
-	data->cmd_data_gpio = device_get_binding(
-			DT_INST_GPIO_LABEL(0, cmd_data_gpios));
-	if (data->cmd_data_gpio == NULL) {
-		LOG_ERR("Could not get GPIO port for cmd/DATA port");
-		return -EPERM;
+	if (!device_is_ready(config->cmd_data_gpio.port)) {
+		LOG_ERR("GPIO port for cmd/DATA port is not ready");
+		return -EINVAL;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:356]:
```suggestion
		return -ENODEV;
```

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/display/display_st7789v.c [hunk 15]
```diff
@@ -263,149 +262,120 @@ static int st7789v_set_orientation(const struct device *dev,
...
 
-	st7789v_transmit(p_st7789v, ST7789V_CMD_RGBCTRL, st7789v_rgb_param,
-			 sizeof(st7789v_rgb_param));
+	st7789v_transmit(dev, ST7789V_CMD_RGBCTRL, config->rgb_param,
+			 sizeof(config->rgb_param));
 }
 
 static int st7789v_init(const struct device *dev)
 {
 	struct st7789v_data *data = (struct st7789v_data *)dev->data;
```
Comment (gmarull) [inline drivers/display/display_st7789v.c:332]:
```suggestion
```
unused

File: drivers/display/display_st7789v.c (hunk 1 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c:111]: @gmarull I will maintain the status-quo in this PR and open a separate PR for the software reset.

Comment (gmarull) [reply drivers/display/display_st7789v.c:111]: I think it's better to address it here (in a separate commit), the only issue with the current status is the `required: True` field in the dt binding.

File: drivers/display/display_st7789v.c (hunk 9 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c:163]: @gmarull After applying the suggestion, the compiler complains about discarding 'const' qualifier. When making all references to `config` const I end up with a last warning for the spi_buf:
```
/home/casper/src/zephyrproject/zephyr/drivers/display/display_st7789v.c: In function 'st7789v_transmit':
/home/casper/src/zephyrproject/zephyr/drivers/display/display_st7789v.c:85:47: warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   85 |         const struct spi_buf tx_buf = {.buf = &cmd, .len = 1};
      |                                               ^
/home/casper/src/zephyrproject/zephyr/drivers/display/display_st7789v.c:93:47: warning: initialization discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
   93 |         const struct spi_buf tx_buf = {.buf = tx_data, .len = tx_count};
      |                                               ^~~~~~~
```
Any ideas how to fix that?

Comment (gmarull) [reply drivers/display/display_st7789v.c:163]: `struct st7789v_config *config` should actually be `const struct st7789v_config *config`

File: drivers/display/display_st7789v.c (hunk 1 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c:111]: Added a commit to fix software reset

File: drivers/display/display_st7789v.c (hunk 9 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c]: I added the `const` modifier to the config struct. However I needed to add a `(void *)` to the tx_buf. I am unsure whether that is the best solution.

File: dts/bindings/display/sitronix,st7789v.yaml [hunk 16]
```diff
@@ -11,7 +11,7 @@ include: spi-device.yaml
 properties:
     reset-gpios:
       type: phandle-array
-      required: true
+      required: false
```
Comment (gmarull) [inline dts/bindings/display/sitronix,st7789v.yaml:14]: nit: removing required is equivalent to false

**Review by gmarull — APPROVED** [review]
Just left a minor comment, but LGTM. Thanks for these improvements!

File: dts/bindings/display/sitronix,st7789v.yaml (hunk 16 above)
Comment (caspermeijn) [reply dts/bindings/display/sitronix,st7789v.yaml:14]: Personally I like that is was explicitly optional and other properties also have `required: false`. Do you want me to remove all `required: false` from this file?

File: drivers/display/display_st7789v.c [hunk 17]
```diff
@@ -63,33 +63,36 @@ struct st7789v_data {
...
 static void st7789v_transmit(const struct device *dev, uint8_t cmd,
-			     uint8_t *tx_data, size_t tx_count)
+			     const uint8_t *tx_data, size_t tx_count)
 {
-	struct st7789v_config *config = (struct st7789v_config *)dev->config;
+	const struct st7789v_config *config = dev->config;
 
-	struct spi_buf tx_buf = { .buf = &cmd, .len = 1 };
-	struct spi_buf_set tx_bufs = { .buffers = &tx_buf, .count = 1 };
+	{
```
Comment (jfischer-no) [inline drivers/display/display_st7789v.c:84]: Why these changes, that bring no benefits at all except more code,  and why are they part of this commit?

File: drivers/display/display_st7789v.c [hunk 18]
```diff
@@ -63,33 +63,36 @@ struct st7789v_data {
...
+	struct st7789v_data *data = dev->data;
 
 	data->x_offset = x_offset;
 	data->y_offset = y_offset;
 }
 
 static void st7789v_set_cmd(const struct device *dev, int is_cmd)
 {
-	struct st7789v_config *config = (struct st7789v_config *)dev->config;
+	const struct st7789v_config *config = dev->config;
```
Comment (jfischer-no) [inline drivers/display/display_st7789v.c:74]: commit https://github.com/zephyrproject-rtos/zephyr/pull/36717/commits/6dbb628b7969fcd74fc011672312c7ce54bff809 introduces st7789v_config, therefore this should be fixed there.

Comment (caspermeijn) [reply drivers/display/display_st7789v.c:74]: I squashed these changes and reordered the commits, so that they make more sense.

File: drivers/display/display_st7789v.c (hunk 17 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c:84]: The order of the commit was a mess, because that was easier then rewrite the history. I fixed that now.

`tx_data` needs to be const, because some of that data now comes from `st7789v_config`. 
`tx_buf` needs to be const, because it holds `tx_data`.
The double code is because you can't write to the const `tx_buf`.
I added a new scope, to prevent shadowing the the `tx_buf` name.
I hope this makes sense.

**Review by gmarull — APPROVED** [review]

**Comment by gmarull** [comment]
@jfischer-no ping

**Review by JordanYates — APPROVED** [review]
SPI changes look fine to me

File: drivers/display/display_st7789v.c (hunk 17 above)
Comment (caspermeijn) [reply drivers/display/display_st7789v.c:84]: I don't understand what you mean. 

If you mean that `tx_data` shouldn't be `const`, then the data from `st7789c_config` can't be transmitted directly. This could be solved by copying the data before calling transmit (as it did in the old code).

If you mean something else, could you explain how to solve your concerns?
